runtime/caml/skiplist.h

```diff
@@ -0,0 +1,102 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*             Xavier Leroy, projet Cambium, INRIA Paris                  */
+/*                                                                        */
+/*   Copyright 2020 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+/* A dictionary data structure implemented as skip lists */
+
+/* Keys and associated data are natural-width integers (type [uintnat]).
+   Pointers can be used too, modulo conversion to [uintnat]. */
+
+#ifndef CAML_SKIPLIST_H
+#define CAML_SKIPLIST_H
+
+#include "config.h"
+
+#define NUM_LEVELS 17
+
+/* The head of a skip list */
+
+struct skiplist {
+  uintnat header1, header2;    /* dummy values for layout compatibility */
+  struct skipcell * forward[NUM_LEVELS]; /* forward chaining */
+  int level;                   /* max level used */
+};
+
+/* The cells of a skip list */
+
+struct skipcell {
+  uintnat key;
+  uintnat data;
+#if (__STDC_VERSION__ >= 199901L)
+  struct skipcell * forward[];  /* variable-length array */
+#else
+  struct skipcell * forward[1]; /* variable-length array */
+#endif
+};
+
+/* Size of struct skipcell, in bytes, without the forward array */
+#if (__STDC_VERSION__ >= 199901L)
+#define SIZEOF_SKIPCELL sizeof(struct skipcell)
+#else
+#define SIZEOF_SKIPCELL (sizeof(struct skipcell) - sizeof(struct skipcell *))
+#endif
```

Comment:
Would this be better either in the C file or definitely guarded with `CAML_INTERNALS`?

Reply:
Definitely!  Moved to the C file.